src/Interpreters/AsynchronousInsertQueue.cpp

```diff
@@ -323,19 +338,24 @@ AsynchronousInsertQueue::pushDataChunk(ASTPtr query, DataChunk chunk, ContextPtr
     {
         std::lock_guard lock(shard.mutex);
 
+        bool elapsed_time_exceeds_max_timeout = false;
         auto [it, inserted] = shard.iterators.try_emplace(key.hash);
         if (inserted)
         {
             auto now = std::chrono::steady_clock::now();
-            auto timeout = now + Milliseconds{key.settings.async_insert_busy_timeout_ms};
-            it->second = shard.queue.emplace(timeout, Container{key, std::make_unique<InsertData>()}).first;
+            if (getContext()->getSettingsRef().allow_experimental_async_insert_adjustable_busy_timeout)
+                updateBusyWaitTimeout(now, shard, elapsed_time_exceeds_max_timeout, shard_num);
```

Comment:
I do not get if it, how it works. We update timeout before we check bytes/queries limit. 
Will we update busy timeout in this case as well? As I understand the algorithm we do not need to batch more if there is enough bytes per batch.

Reply:
We do not mark the whole queue shard for immediate processing, but rather only the entries with the specific `key.hash` that have reached either bytes or # of queries limit:

https://github.com/ClickHouse/ClickHouse/blob/06274c7bc641b10038a696549e384b756654ccea/src/Interpreters/AsynchronousInsertQueue.cpp#L326-L354

Other entries that have fallen into the same queue shard will remain in the async queue. 
Hence, I believe, we should adjust the timeout.